pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceAssignmentDriver.java

```diff
@@ -51,16 +52,34 @@ public InstanceAssignmentDriver(TableConfig tableConfig) {
     _tableConfig = tableConfig;
   }
 
+  /**
+   * Assign instances to InstancePartitions object.
+   * @param instancePartitionsType type of instance partitions
+   * @param instanceConfigs list of instance configs
+   * @param partitionToInstancesMap existing instance with sequence that should be respected. An empty list
+   *                                      means no preceding sequence to respect and the instances would be sorted.
+   */
   public InstancePartitions assignInstances(InstancePartitionsType instancePartitionsType,
-      List<InstanceConfig> instanceConfigs) {
+      List<InstanceConfig> instanceConfigs, Map<Integer, List<String>> partitionToInstancesMap) {
+    boolean shouldRetainInstanceSequence = !partitionToInstancesMap.isEmpty();
     String tableNameWithType = _tableConfig.getTableName();
-    LOGGER.info("Starting {} instance assignment for table: {}", instancePartitionsType, tableNameWithType);
+    LOGGER.info("Starting {} instance assignment for table: {}. Should retain instance sequence: {}",
+        instancePartitionsType, tableNameWithType, shouldRetainInstanceSequence);
 
     InstanceAssignmentConfig assignmentConfig =
         InstanceAssignmentConfigUtils.getInstanceAssignmentConfig(_tableConfig, instancePartitionsType);
     InstanceTagPoolSelector tagPoolSelector =
         new InstanceTagPoolSelector(assignmentConfig.getTagPoolConfig(), tableNameWithType);
-    Map<Integer, List<InstanceConfig>> poolToInstanceConfigsMap = tagPoolSelector.selectInstances(instanceConfigs);
+    Map<Integer, List<InstanceConfig>> poolToInstanceConfigsMap =
+        tagPoolSelector.selectInstances(instanceConfigs, partitionToInstancesMap);
+
+    InstancePartitions instancePartitions = new InstancePartitions(
+        instancePartitionsType.getInstancePartitionsName(TableNameBuilder.extractRawTableName(tableNameWithType)));
+    if (shouldRetainInstanceSequence) {
+      // Keep the partition to instances map if instance sequence should be retained.
+      instancePartitions
+          .setPartitionToInstancesMap(extractInstanceNamesFromPoolToInstanceConfigsMap(poolToInstanceConfigsMap));
+    }
 
     InstanceConstraintConfig constraintConfig = assignmentConfig.getConstraintConfig();
```

Comment:
Did we think about wrapping this under a new concrete implementation of `InstanceConstraintApplier`

The current constraint applier rotates the instances to alleviate hotspots but that happens after the `tagPoolSelector` has already selected the candidate instances and sorted them. 

Can handle the requirement in this PR by moving the constraint applier inside the `tagPoolSelector` ?

Reply:
Related question -- should we create a new strategy for this to simplify code and avoid adding multiple conditions to existing code ?

Reply:
Yes I did think of extending the `InstanceConstraintApplier`, but I do want to keep the current logic if the `existingPoolToInstancesMap` is not given. Plus, it'd be better to store the instance with the sequence before the rotation, as it will always return the deterministic list of instance if the input instance sequence is fixed. And I don't like the idea of overriding the existing method as it'd be hard to maintain two copies of code if some changes is needed.

Reply:
IMO the code will be much cleaner and maintainable if we create a new strategy -- for example we can create a new `InstanceTagPoolSelector` at the cost of duplicating some code that stays the same.

Driver can probably stay the same. 

Thoughts ? cc @jasperjiaguo